### ISSUE TITLE
Are these the changes needed to reproduce inductive unsupervised learning with Infomax?

### ISSUE THREAD

**Issue by dxcl** [rusty1s/pytorch_geometric#1289]
## ❓ Questions & Help

Hi! This is a fantastic piece of kit and I'm currently using the Infomax model for unsupervised embedding generation of the PPI data. I was just checking your thoughts on modifying the example given to use it for this task.

The paper (see section 4.2) refers to the PPI model involving a "three-layer mean-pooling model with dense skip connections", and from what I can see it seems to stack GraphSAGE layers with output length = embedding_length (but I'm wondering whether I need to add something else). As well as this, they use a different corruption function, by sampling another graph from the training set and applying dropout to it.

So if I'm trying to reproduce the results of the original paper, I was thinking something like this might work:

```
class Encoder(nn.Module):
    def __init__(self, in_channels, hidden_channels):
        super(Encoder, self).__init__()
        self.conv1 = SAGEConv(in_channels, hidden_channels)
        self.conv2 = SAGEConv(hidden_channels, hidden_channels)
        self.conv3 = SAGEConv(hidden_channels, hidden_channels)
        self.prelu = nn.PReLU(hidden_channels)

    def forward(self, x, edge_index):
        x = self.conv1(x, edge_index)
        x = self.prelu(x)
        x = self.conv2(x, edge_index)
        x = self.prelu(x)
        x = self.conv3(x, edge_index)
        x = self.prelu(x)
        return x

def corruption(x, edge_index):
    for sampled in train_loader.__iter__():
        if x.shape != sampled.x.shape:
            sampled = sampled.to(device)
            return d = nn.Dropout(p=0.5)
            return d(sampled.x), sampled.edge_index
        return x, edge_index

def train():
    model.train()
    total_loss = 0
    for data in train_loader:
        num_graphs = data.num_graphs
        data.batch = None
        data = data.to(device)
        optimizer.zero_grad()
        pos_z, neg_z, summary = model(data.x, data.edge_index)
        loss = model.loss(pos_z, neg_z, summary)
        total_loss += loss.item() * num_graphs
        loss.backward()
        optimizer.step()
    return total_loss / len(train_loader.dataset)
```

When I run this, the model's loss jumps all over the place, but it definitely drops. However, when I then try to fit a Logistic Regression model to the embeddings, to get the F1 scores for predicting the labels, I can't get an F1 above 0.48, whilst the paper has a result of 0.64.

I'm collecting embeddings like this:
```
def collect_outputs(loader):
    model.eval()
    ys, preds = [], []
    for data in loader:
        ys.append(data.y)
        z, _, _ = model(data.x.to(device), data.edge_index.to(device))
        preds.append(z)
    return ys, preds
```

Thoughts on how to replicate the results? I reckon I might be missing something...

**Comment by rusty1s** [comment]
Is the `return` statement a mistake here?
```
return d = nn.Dropout(p=0.5)
```

Otherwise, this looks quite good to me, although the paper mentions another `Encoder` architecture, *e.g.*, it reuses node embeddings in the first layer for the last layer. Otherwise, there may be some differences in hyperparameter settings, so please verify that they are the same.

**Comment by dxcl** [comment]
Ah yes, there shouldn't be a return statement in that line. Will have a look at how the dense skip connections are implemented in the Encoder and review the hyperparameters

**Comment by dxcl** [comment]
Happily this was fixed by adding the skip connections (i.e.
```python
x1 = self.conv1(x0 + xw, edge_index)
x1 = self.prelu1(x1)
````
etc.), and (more importantly) using a MultiOutputRegressor on the resulting embeddings.
